File: modules/ia_cardlink/android/src/main/kotlin/de/ihreapotheken/appsdk/flutter/card_link/IaSdkFlutterCardLink.kt
package de.ihreapotheken.appsdk.flutter.card_link

import android.app.Activity
import de.ihreapotheken.appsdk.flutter.card_link.events.CardLinkEventSender
import de.ihreapotheken.appsdk.flutter.card_link.methods.DeleteCardHandler
import de.ihreapotheken.appsdk.flutter.card_link.methods.GetSavedCardsHandler
import de.ihreapotheken.appsdk.flutter.card_link.methods.LaunchHandler
import de.ihreapotheken.sdk.cardlink.CardlinkModule
import de.ihreapotheken.sdk.cardlink.api.CardLink
import de.ihreapotheken.sdk.cardlink.api.REQUEST_CODE_CARD_LINK_ACTIVITY
import de.ihreapotheken.sdk.integrations.api.IaSdk
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** IaSdkFlutterCardLink */
class IaSdkFlutterCardLink :
    FlutterPlugin,
    MethodCallHandler,
    ActivityAware {

    private lateinit var channel: MethodChannel
    private var activity: Activity? = null

    private val eventSender by lazy {
        CardLinkEventSender(
            activityProvider = { activity },
            methodChannelProvider = { channel }
        )
    }

    private val launchHandler by lazy {
        LaunchHandler(
            activityProvider = { activity },
            eventSender = eventSender
        )
    }

    private val getSavedCardsHandler by lazy {
        GetSavedCardsHandler(activityProvider = { activity })
    }

    private val deleteCardHandler by lazy {
        DeleteCardHandler(activityProvider = { activity })
    }

    companion object {
        private const val CHANNEL_ID = "de.ihreapotheken/sdk/cardLink"
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(
            flutterPluginBinding.binaryMessenger,
            CHANNEL_ID
        )
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addActivityResultListener { requestCode, resultCode, data ->
            if (requestCode == REQUEST_CODE_CARD_LINK_ACTIVITY) {
                if (resultCode == Activity.RESULT_OK) {
                    val message = data?.getStringExtra("card_link_result")
                    eventSender.sendCardLinkResultEvent(message)
                } else if (resultCode == Activity.RESULT_CANCELED) {
                    eventSender.sendCardLinkCanceledEvent()
                }
                true
            } else {
                false
            }
        }
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity()
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "register" -> {
                IaSdk.register(CardlinkModule)
                result.success(null)
            }

            "launch" -> {
                // TODO: Pass coreAppLogFileURL to CardLink when supported on Android
                launchHandler.handle(call, result)
            }

            "getVersion" -> {
                result.success(null)
            }

            "getEnvironment" -> {
                result.success(null)
            }

            "getLogFilePath" -> {
                activity?.let {
                    val logFilePath = CardLink.getLogFilePath(it)
                    result.success(logFilePath)
                } ?: result.success("")
            }

            "getSavedCards" -> {
                getSavedCardsHandler.handle(call, result)
            }

            "deleteCard" -> {
                deleteCardHandler.handle(call, result)
            }

            "deleteAllCards" -> {
                activity?.let {
                    CardLink.clearAllCardLinkData(it)
                    result.success(null)
                } ?: result.error("NO_ACTIVITY", "Activity not available", null)
            }

            "deleteAllUserRelatedData" -> {
                activity?.let {
                    CardLink.clearAllCardLinkData(it)
                    result.success(null)
                } ?: result.error("NO_ACTIVITY", "Activity not available", null)
            }

            "finish" -> {
                result.success(null)
            }

            else -> {
                result.notImplemented()
            }
        }
    }
}

File: modules/ia_cardlink/ios/ia_cardlink/Sources/ia_cardlink/IaSdkFlutterCardLink.swift
import Flutter
import UIKit
import IACore
import IACardLink

@MainActor
public class IaSdkFlutterCardLink: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel?

    private let eventSender = CardLinkEventSender()
    private lazy var launchHandler = LaunchHandler(eventSender: eventSender)
    private let getSavedCardsHandler = GetSavedCardsHandler()
    private let deleteCardHandler = DeleteCardHandler()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "de.ihreapotheken/sdk/cardLink",
            binaryMessenger: registrar.messenger()
        )
        let instance = IaSdkFlutterCardLink()
        registrar.addMethodCallDelegate(instance, channel: channel)
        instance.channel = channel
        instance.eventSender.setMethodChannel(channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "register":
            IASDK.register([IASDKModule.cardLink])
            result(nil)

        case "launch":
            guard let args = call.arguments as? [String: Any] else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments received", details: nil))
                return
            }
            launchHandler.handle(args: args, result: result)

        case "getVersion":
            result(CardLink.version)

        case "getEnvironment":
            result(CardLink.environment.pluginStringValue)

        case "getLogFilePath":
            result(CardLink.logsPath)

        case "getSavedCards":
            guard let args = call.arguments as? [String: Any] else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments received", details: nil))
                return
            }
            getSavedCardsHandler.handle(args: args, result: result)

        case "deleteCard":
            guard let args = call.arguments as? [String: Any] else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments received", details: nil))
                return
            }
            deleteCardHandler.handle(args: args, result: result)

        case "deleteAllCards":
            do {
                let status = try CardLink.deleteAllCards()
                result("\(status)")
            } catch {
                result(FlutterError(code: "DELETE_ALL_CARDS_ERROR", message: "\(error)", details: nil))
            }

        case "deleteAllUserRelatedData":
            Task {
                do {
                    try await CardLink.deleteAllUserRelatedData()
                    result(nil)
                } catch {
                    result(FlutterError(code: "DELETE_ALL_USER_DATA_ERROR", message: "\(error)", details: nil))
                }
            }

        case "finish":
            CardLink.finish {
                result(nil)
            }

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}
